modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/service/indexer/cache/KeyValueStoreSubscriber.scala

```diff
@@ -0,0 +1,111 @@
+/**
+  * A subscriber actor that receives messages from the key value store whenever a change occurs.
+  *
+  * @param onChange the method that gets triggered when a change to key value store occurs
+  * @tparam K the key type
+  * @tparam V the value type
+  */
+class KeyValueStoreSubscriber[K, V] private (onChange: OnKeyValueStoreChange[K, V]) extends Actor {
+
+  private val key      = LWWMapKey[K, V](self.path.name)
+  private var previous = Map.empty[K, V]
+
+  private def diff(recent: Map[K, V]): KeyValeStoreChanges[K, V] = {
+    val added   = (recent -- previous.keySet).map { case (k, v) => ValueAdded(k, v) }.toSet
+    val removed = (previous -- recent.keySet).map { case (k, v) => ValueRemoved(k, v) }.toSet
+
+    val modified = (recent -- added.map(_.key)).foldLeft(Set.empty[KeyValeStoreChange]) {
+      case (acc, (k, v)) =>
+        previous.get(k).filter(_ == v) match {
+          case None => acc + ValueModified(k, v)
+          case _    => acc
+        }
+    }
+    KeyValeStoreChanges(added ++ modified ++ removed)
+  }
+
+  override def receive: Receive = {
+    case c @ Changed(`key`) =>
+      val map     = c.get(key).entries
+      val changes = diff(map)
+      if (changes.values.nonEmpty) onChange(changes)
+      previous = map
+
+    case _ => // drop
+  }
+}
+
+object KeyValueStoreSubscriber {
+
+  /**
+    * Enumeration of types related to changes to the key value store.
+    */
+  sealed trait KeyValeStoreChange extends Product with Serializable
+  object KeyValeStoreChange {
+
+    /**
+      * Signals that an element has been added to the key value store.
+      *
+      * @param key   the key
+      * @param value the value
+      */
+    final case class ValueAdded[K, V](key: K, value: V) extends KeyValeStoreChange
+
+    /**
+      * Signals that an already existing element has been updated from the key value store.
+      *
+      * @param key   the key
+      * @param value the value
+      */
+    final case class ValueModified[K, V](key: K, value: V) extends KeyValeStoreChange
+
+    /**
+      * Signals that an already existing element has been removed from the key value store.
+      *
+      * @param key   the key
+      * @param value the value
+      */
+    final case class ValueRemoved[K, V](key: K, value: V) extends KeyValeStoreChange
+  }
+
+  /**
+    * The set of changes that have occurred on the primary store
+    *
+    * @param values the set of changes
+    */
+  final case class KeyValeStoreChanges[K, V](values: Set[KeyValeStoreChange])
+
+  /**
+    * Constructs the [[KeyValueStoreSubscriber]] actor.
+    *
+    * @param id       the ddata key
+    * @param onChange the method that gets triggered whenever a change to the key value store occurs
+    * @tparam K the key type
+    * @tparam V the value type
+    * @return an [[ActorRef]] of the [[KeyValueStoreSubscriber]] actor
+    */
+  final def apply[K, V](id: String, onChange: OnKeyValueStoreChange[K, V])(implicit as: ActorSystem): ActorRef =
```

Comment:
the fact that the actor id is equal to the map key it means we can only have a single actor per cache. we could instead pass the id to the actor constructor, let it subscribe by itself and leave the actor id to be automatically generated.

Reply:
this should allow us to have multiple subscriptions on the store api, something like:
`def subscribe(f: KeyValueStoreChanges[K, V] => ()): F[Subscription] = ???`

and
```
class Subscription[F[_]] private(key: String, ref: ActorRef) {
  def unsubscribe: F[Unit] = ???
}
```

Reply:
One actor per cache makes sense and it is more than enough. We are not planning on having 300 caches.